### PR TITLE
Use union types for identifiers

### PR DESCRIPTION
**Pull request by mavnn**
This style recommendation is still up for debate: opinions and thoughts are welcomed!

### CODE REVIEW

**Review by juanedi — COMMENTED** [review]
This looks great!

I wonder if we should talk about this in a more general setting, since using custom types for domain concepts may be useful for things other than ids (I'm thinking about urls and tokens but it could be anything).

I think it would be great if we could frame the suggestion in a more general way, or at least add a comment at the end in the lines of "note that using this technique might be useful in other contexts too".

File: README.md [hunk 1]
```diff
@@ -348,6 +349,72 @@ raw JSON.
...
+getStudentDetails sid =
+    -- Passing a name in here by mistake will compile fine --
+    ...
+```
+
+```elm
+-- Do this instead --
+type alias Student =
+    { id : StudentId
+    , name : String
```
Comment (jwoudenberg) [inline README.md]: Should we in a similar fashion encourage this to be a `StudentName`, and age a `StudentAge`?

Comment (mavnn) [reply README.md]: That is a very good (although maybe orthogonal?) question. I mean, the broader thinking here is that anything which has a domain specific meaning should have a domain specific type: but the flip side is that StudentName has very few constraints and you're likely to do a lot of string operations on it.

If you can think of a reasonably clear set of guidelines for where to switch to a type, I'm all for it!

Comment (jwoudenberg) [reply README.md]: Yeah, sorry, this is definitely orthogonal.

> If you can think of a reasonably clear set of guidelines for where to switch to a type, I'm all for it!

That's a fair point! I don't know If I can come up with such a guideline. Maybe a process to default to a type and remove it if it turns out to be really impractical.

Either way not something we need to pull into this PR :).

**EDIT**: Also noticed Juan already made the exact same point in his review.

Comment (mavnn) [reply README.md]: That's one of the things making me wonder if it really is orthogonal. On the other hand, I've seen the ID specific application catch enough bugs I wouldn't want to lose calling it out.

File: README.md [hunk 2]
```diff
@@ -348,6 +349,72 @@ raw JSON.
...
+    , age : Int
+    }
+
+
+type StudentId
+    = StudentId String
```
Comment (avh4) [inline README.md]: Maybe we should prefer to call this `Id`, so the normal use will be

```elm
import My.Student as Student exposing (Student)

type alias MyType =
   { students : Dict Student.Id Student
   }
```

**Comment by jamesgary** [comment]
I like this!

If we want to unwrap the string value from `StudentId` to, say, build a url for an ajax call, should we make a `toUrl : StudentId -> Url` function in the `Student` module, or expose `toString : StudentId -> String` and use that whenever we need the string? I feel like the former is preferable, but I'm wondering if it's ever desirable to expose `toString` unwrapping function?

File: README.md (hunk 2 above)
Comment (mavnn) [reply README.md]: Putting my contrary hat on... how would this work for types where there are multiple IDs in play? If I'm creating a `Model` for a bank transfer page:

```elm
type alias Model =
    { from : Customer.Id
    , fromAccount : Account.Id
    , to : Customer.Id
    , toAccount : Account.Id
    }
```

Do I now have to create separate files/modules for Account and Customer?

**Comment by mavnn** [comment]
Personally I'd lean towards the IDs being opaque types and keep all conversions within the same module. But that's probably a case by case call.

**Review by jwoudenberg — APPROVED** [review]
I like this change! There's many exciting ways this can be taken further, but I would already be happy merging this as is!

File: README.md (hunk 2 above)
Comment (rtfeldman) [reply README.md]: I think the bigger problem here is that I think it's best if the identifier goes in its own module, e.g. https://github.com/rtfeldman/elm-spa-example/blob/master/src/CommentId.elm

This way we can expose only a `Decoder` for it and not a `fromString` or `fromInt` or anything like that, making it really clear that this thing should only ever come from the server. We should never make one by hand.

In that world, there's a significant difference between `module Customer.Id` and `module CustomerId`. Compare:


```elm
import Customer
import CustomerId exposing (CustomerId)

myId : CustomerId
```

```elm
import Customer
import Customer.Id

myId : Customer.Id.Id
```

If we want to annotate it as `Customer.Id` while already having a `Customer` module imported, then we have to do something like this:

```elm
import Customer
import Customer.Id as Customer

myId : Customer.Id
```

That solves the immediate problem, but creates a new one: now `Customer.decoder` is ambiguous. Does it mean the `decoder` from the `Customer` module or the one from the aliased `Customer.Id` module?

tl;dr I think `module CustomerId` is the way to go here. 😄

Comment (stoeffel) [reply README.md]: > Do I now have to create separate files/modules for Account and Customer?

Yes, I would actually prefer that over a union for an id in the same file, I would like that to be the default.

**Comment by rtfeldman** [comment]
> Personally I'd lean towards the IDs being opaque types and keep all conversions within the same module. But that's probably a case by case call.

I'd go as far as to say we should make this part of the style guide. Specifically I'd propose these style guide additions:

1. Whenever we have an identifier like `CustomerId` or `Username`, make it an opaque type in its own module.
2. The module name should be the same as the identifier name, and should not be further qualified. (For example, use `PracticeId`, not `Nri.Quiz.PracticeId`. If we did the latter, we would just end up doing `import Nri.Quiz.PracticeId as PracticeId` in practice anyway, introducing an unnecessary amount of indirection when figuring out which values came from which modules.)
4. The module should expose an intentionally minimal set of conversion functions.
    * We should never expose conversion functions before they're needed. Start with `module CustomerId exposing (CustomerId)` and expand the API from there only as necessary!
    * We should avoid exposing conversion functions whose types depend on the internals of the custom type. (For example, `decoder : Decoder CustomerId` is great, but functions like `fromInt : Int -> CustomerId` make the system brittle to implementation details. Exposing an `Int -> CustomerId` function should be avoided unless there is a very important *production* reason to introduce it. If tests need an `Int -> CustomerId` function, they can easily write one using `CustomerId.decoder` and `Debug.crash`, since in tests `Debug.crash` is as harmless as any other test failure.)

File: README.md (hunk 1 above)
Comment (rtfeldman) [reply README.md]: One guideline is "might the underlying implementation change?"

For example, we might conceivably switch from integers to strings for database identifiers. (It's very unlikely, but conceivable nonetheless.)

`name` is always going to be a string though.

**Comment by mavnn** [comment]
I'm going to disagree with "2" above with my build engineer hat on: I don't think we should make our tooling/linting more complex purely to slightly reduce typing, and there might be cases where we have multiple similarly named IDs floating around.

I'll wrap up this weeks Priority of the week by adding impression of the consensus above 😄
